bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManagerImpl.java

```diff
@@ -1116,6 +1143,38 @@ private void setThingStatus(Thing thing, ThingStatusInfo thingStatusInfo) {
         }
     }
 
+    @Override
+    public void setEnabled(ThingUID thingUID, boolean enabled) {
+        Thing thing = getThing(thingUID);
+
+        if (enabled) {
+            // Enable a thing
+            logger.debug("Thing {} will be enabled.", thing.getUID().getAsString());
+            if (storage != null) {
+                storage.remove(thingUID.getAsString());
+            }
+            if (!ThingHandlerHelper.isHandlerInitialized(thing)) {
```

Comment:
You didn't answer [my question](https://github.com/eclipse/smarthome/pull/5191#discussion_r201306064) about the purpose if this if-statement but just changed it to this. 

I assume you want to prevent initializing an already existing thing handler for the second time? 

If that's the case and this really _could_ happen, then I see a problem here: `isHandlerInitialized(...)` only tells you if the initialization phase of a potentially existing handler is _finished_  or not. It doesn't tell you if there already exists a handler which might still be INITIALIZING. Therefore this will lead to awkward race conditions, at worst leaving two handlers working on the same device. For that matter, `isHandlerRegistered(...)` (which you have used below too) would have been the better choice, I think.

On the other hand, if you follow the code flow a little, you will see in `registerHandler(...)` that the code already _is_ protected against such a race-condition itself. Also, `initializeHandler(...)` is protected against such a scenario. Leaving ugly warnings in the log though - so it might still be good to do have this kind of protection here too. 

Nevertheless, the most "dangerous" part in this context is setting the thing status to `UNINITIALIZED`/`NONE` blindly, regardless of its previous status. Don't you think it would be better to check for the previous status first and make sure that it _really_ is `UNINITIALIZED`/`DISABLED` and _only then_ change it to `UNINITIALIZED`/`NONE` and call `registerAndInitializeHandler(...)`?

Reply:
My assumption here was that, when we call `setEnabled(thingUID, true)` we want to bring the thing to `ONLINE` status regardless of its previous status. That is why the condition here was `if (!thing.getStatus().equals(ThingStatus.ONLINE))` before my last commit. So, besides your comments on the code, does this sounds like the correct approach to you?